poly/joint.go

```diff
@@ -0,0 +1,284 @@
+
+	// the Receiver private / public key combination
+	// it may or may not have to be the long term key of the node
+	Key *config.KeyPair
+
+	// List of Dealers. Be careful : this receiver should have the SAME index for all the Dealer's promises !!
+	// otherwise we wouldn't know which index to chose from the shared public polynomial
+	Dealers []*Dealer
+
+	// When the dealers are all done, we can compute the shared secret which consists of a
+	// 1. Public Polynomial which is basically the sums of all Dealers's polynomial
+	// 2. Share of the global Private Polynomial (which is to never be computed directly), which is
+	// 		basically SUM of fj(i) for a receiver i
+	Secret SharedSecret
+}
+
+// NewDealer returns a newly created & intialized Dealer struct
+func NewDealer(info PolyInfo, secret, promiser *config.KeyPair, receiverList []abstract.Point) *Dealer {
+	return new(Dealer).Init(info, secret, promiser, receiverList)
+}
+
+// Dealer.Init inits a new Dealer structure :
+// That basically create the promise of the dealer and the respective shares using the list of receivers
+func (d *Dealer) Init(info PolyInfo, secret, promiser *config.KeyPair, receiverList []abstract.Point) *Dealer {
+	d.Info = info
+	d.Promise = new(Promise).ConstructPromise(secret, promiser, info.T, info.R, receiverList)
+	d.State = new(State).Init(*d.Promise)
+	return d
+}
+
+// Basically a wrapper around Promise / Response so that a dealer can verify that all its receiver correctly received its promise and are not cheating
+func (d *Dealer) AddResponse(i int, response *Response) error {
+	return d.State.AddResponse(i, response)
+}
+
+// A wrapper around State.PromiseCertified for this dealer. It must have received enough Response (and/or max number of blameProof)
+func (d *Dealer) Certified() error {
+	return d.State.PromiseCertified()
+}
+
+func NewReceiver(info PolyInfo, key *config.KeyPair) *Receiver {
+	return new(Receiver).Init(info, key)
+}
+
+// Init a new Receiver struct
+// info is the info about the structure of the polynomials used
+// key is the long-term public key of the receiver
+func (r *Receiver) Init(info PolyInfo, key *config.KeyPair) *Receiver {
+	r.index = -1 // no dealer received yet
+	r.info = info
+	r.Key = key
+	r.Dealers = make([]*Dealer, 0, info.N)
+	return r
+}
+
+// AddDealer adds a dealer to the array of dealers the receiver already has.
+// You must give the index of the receiver in the promise of the dealer,
+// i.e. index is generally the index of the receiver in the matrix, and
+// is usually fixed.
+// Most of the time it will be like , for peer i :
+// receiver(i).AddDealer(i,dealer(j)) for all j
+// It will return a Response to be sent back to the Dealer so he can verify its promise
+func (r *Receiver) AddDealer(index int, dealer *Dealer) (*Response, error) {
+	if r.index == -1 {
+		r.index = index
+	}
+	if r.index != index {
+		return nil, errors.New(fmt.Sprintf("Wrong index received for receiver : %d instead of %d", index, r.index))
```

Comment:
Is this an error that should ever be able to happen outside the control of this program (due to external circumstances such as some other node's action or an incorrect message being received)?  If not - i.e., if this error can only happen because of a program bug in the caller of this function - then this should be a panic, not an error return.  And if there's *no* way this function should ever be able to fail due to external, uncontrollable circumstances (only program bugs), then it probably shouldn't have an error return at all.